File: Anagrams.cpp
class Solution {
	public:
		vector<string> anagrams(vector<string> &strs) {
			vector<string> res;
			map<string ,int> anagrams;
			
			//这里关键是在int中存入第一个出现的下标，而不是次数。
			//以后出现可以利用find来判断，不需要记录次数
			for(int i = 0; i != strs.size(); i++)
			{
				string s = strs[i];
				sort(s.begin(), s.end());
				auto l = anagrams.find(s);
				if(l == anagrams.end()){
					anagrams[s] = i;
				}else{
					res.push_back(strs[i]);
					if(l->second != -1)
					{
						res.push_back(strs[l->second]);
						l->second = -1;
					}
				}
			}

			return res;
		}
};

File: lintcode/datastructure/Anagrams.py
class Solution:
    """
    @param strs: A list of strings
    @return: A list of strings
    """
    def anagrams(self, strs):
        # write your code here
        from collections import defaultdict
        res = []
        mp = defaultdict(lambda: 0)
        for each in strs:
            mp[''.join(sorted(each))] += 1
        for each in strs:
            if mp[''.join(sorted(each))] > 1:
                res.append(each)
        return res